src/lib/UXDController.ts

```diff
@@ -0,0 +1,206 @@
+
+export interface CollateralInfo {
+  symbol?: string;
+  token: string;
+  redeemable: BigNumber;
+  minted: BigNumber;
+}
+
+export class UXDController {
+  protected provider: providers.Provider;
+
+  // internal contracts
+  protected controllerContract: UXDControllerContract;
+  protected uxdContract: ERC20Contract;
+
+  // clients can listen to events on these subjects
+  public readonly mintSubject: Subject<MintedEventObject> = new Subject<MintedEventObject>();
+  public readonly redeemSubject: Subject<RedeemedEventObject> = new Subject<RedeemedEventObject>();
+  public readonly uxdApprovalSubject: Subject<ApprovalEventObject> = new Subject<ApprovalEventObject>();
+  public readonly uxdTransferSubject: Subject<TransferEventObject> = new Subject<TransferEventObject>();
+
+  constructor({
+    provider,
+    controllerAddress,
+    uxdTokenAddress,
+  }: {
+    provider: providers.Provider;
+    controllerAddress: string;
+    uxdTokenAddress: string;
+  }) {
+    this.provider = provider;
+
+    this.controllerContract = UXDController__factory.connect(controllerAddress, this.provider);
+    this.uxdContract = ERC20__factory.connect(uxdTokenAddress, this.provider);
+
+    this.registerEventListeners();
+  }
+
+  public mint({
+    market,
+    collateral,
+    ethAmount,
+    slippage,
+    signer,
+  }: {
+    market: string;
+    collateral: string;
+    ethAmount: BigNumber;
+    slippage: BigNumber;
+    signer: Signer;
+  }): Promise<ContractTransaction> {
+    return this.controllerContract.connect(signer).mint(
+      market,
+      collateral,
+      ethAmount,
+      slippage
+    );
+  }
+
+  public redeem({
+    market,
+    collateral,
+    uxdAmount,
+    slippage,
+    signer,
+  }: {
+    market: string;
+    collateral: string;
+    uxdAmount: BigNumber;
+    slippage: BigNumber;
+    signer: Signer;
+  }): Promise<ContractTransaction> {
+    return this.controllerContract.connect(signer).redeem(
+      market,
+      collateral,
+      uxdAmount,
+      slippage
+    );
+  }
+
+  public mintWithEth({
+    market,
+    ethAmount,
+    slippage,
+    signer,
+  }: {
+    market: string;
+    ethAmount: BigNumber;
+    slippage: BigNumber;
+    signer: Signer;
+  }): Promise<ContractTransaction> {
+    return this.controllerContract.connect(signer).mintWithEth(
+      market,
+      slippage,
+      {
+        value: ethAmount,
+      },
+    );
+  }
+
+  public redeemEth({
+    market,
+    uxdAmount,
+    slippage,
+    signer,
+  }: {
+    market: string;
+    uxdAmount: BigNumber;
+    slippage: BigNumber;
+    signer: Signer;
+  }): Promise<ContractTransaction> {
+    return this.controllerContract.connect(signer).redeemEth(market, uxdAmount, slippage);
+  }
+
+  // ???Rework returned infos???
+  public async getCollateralInfo(): Promise<UXDControllerNamespace.CollateralInfoStructOutput[]> {
+    return this.controllerContract.getCollateralInfo();
+  }
+
+  public approveUXD({
```

Comment:
move to `approveToken` ? In which cases do we need this `approve` methods ?

Reply:
This is a convenience method since we deal mostly with UXD. The other method can be used to approve any ERC-20, but that's actually not really needed in the library